Formula/adguard-home.rb

```diff
@@ -0,0 +1,114 @@
+class AdguardHome < Formula
+  desc "Network-wide ads & trackers blocking DNS server"
+  homepage "https://adguard.com/adguard-home.html"
+  url "https://github.com/AdguardTeam/AdGuardHome/archive/v0.98.1.tar.gz"
+  sha256 "37779d659bb9faa1b2273f70aba4db14a77b4467888953d055547f28e8afa707"
+  head "https://github.com/AdguardTeam/AdGuardHome.git", :using => :git
+
+  depends_on "go" => :build
+  depends_on "node@8" => :build
```

Comment:
Does it not work with the newest node version? Because in that case it might not be a good fit for homebrew-core since these versioned formula are mostly there as a stopgap until all software supports the latest version (which is not good enough for new additions)

Reply:
Node.js is only used to build the UI. It works with Node.js v10, should work with v12.

Reply:
Please just use `depends on "node"` then. In addition to that it's a much better experience not to use packer and just depend on node all the time, we're shipping it anyway and it won't risk shipping random outdated node versions with formulae.

Reply:
Agreed. Please use the latest `nodejs` unless you can make a `test do` block that fails with it and please don't use `packer`.

Reply:
I can update the `node` requirement, but `packer` is required to embed the built UI artefacts inside the go binary.

Honestly, I'm a bit tired with all this ping-pong around a simple PR. It's demotivating.

Reply:
> I can update the `node` requirement

Thanks.



> but `packer` is required to embed the built UI artefacts inside the go binary.

Can you elaborate on this, please? Does this use `node` at runtime at all? What is "the UI"?



> Honestly, I'm a bit tired with all this ping-pong around a simple PR.

It's not a simple PR.



> I wanted to share a formula that I'm using, but you clearly either want perfection (while there's tons of formula not following such rules)

New formulae have higher standards than existing formulae. We do want new formula to be as good as possible before they are merged, yes. We do want this merged.

Reminder that you're asking volunteers in their spare time to work with you to get your PR up to their standards. In some cases (e.g. caveats) it's taken longer than it needs to because your first response to being asked to change them was to argue.

We all would like to get this PR merged. Once it is merged, though, we need to be confident we have sufficient understanding of the code and underlying application to be able to maintain it in future. If that's too much work for you: that's fine, we cannot make any demands of you to do this, but it is what we need to be able to run this package manager at the scale we do.

Reply:
> Can you elaborate on this, please? Does this use node at runtime at all? What is "the UI"?

As I've already said, `node` is only used during the build step to build the UI assets. "The UI" is a web application used for configuring and managing server part of AdGuard Home.

> because your first response to being asked to change them was to argue.

At what point? If you see asking questions or asking for examples as "arguing" then I really don't know what to say in response 🤷‍♂ 

> Reminder that you're asking volunteers in their spare time

Reminder, I'm a volunteer doing this in my spare time. I'm not related to homebrew team, nor the AdGuard team. I'm not even a maintainer of any of such projects. Do you really want to use _goodwill_ and _spare time_ as arguments in discussion then? 😉 I see that this community is not very welcoming. Let's put all of this unnecessary discussion aside then, and just focus on this PR.

Current state:
- using `node` doesn't work, because installer doesn't have access to `npm` in the build step
- `node@8` works, `node@10` works
- `packer` is required to embed the built UI into the go binary

Reply:
> As I've already said, `node` is only used during the build step to build the UI assets. 

You said this. You did not state explicitly that it was not used in any form at runtime.

> "The UI" is a web application used for configuring and managing server part of AdGuard Home.

You did not say this before.

> If you see asking questions or asking for examples as "arguing" then I really don't know what to say in response 🤷‍♂

https://github.com/Homebrew/homebrew-core/pull/44051#discussion_r340518319
https://github.com/Homebrew/homebrew-core/pull/44051#discussion_r331395595

> Do you really want to use _goodwill_ and _spare time_ as arguments in discussion then? 😉

Yes, I do. The Homebrew maintainers spend a lot more time (collectively and individually) helping people contribute to this project. 

> I see that this community is not very welcoming. 

I'm not sure most of the >7000 people who have contributed to this repository would agree. Anecdotally, Homebrew has been cited multiple times as being particularly welcoming. In PRs like this on some other projects there would be no comments and this PR would sit unmerged forever rather than having multiple maintainers working to try and work with you to get this to a point where it can be merged. 

> Let's put all of this unnecessary discussion aside then, and just focus on this PR.

"Honestly, I'm a bit tired with all this ping-pong around a simple PR. It's demotivating." were your comments on this PR when previously all comments were technical. Moving onto the technical (which I will happily make the rest of my comments on this PR if you can do the same).

> * using `node` doesn't work, because installer doesn't have access to `npm` in the build step
> * `node@8` works, `node@10` works

Which installer? The `node` formula installs NPM to it's unclear why this would be different. What do you mean "doesn't work"? What is the failure?

> * `packer` is required to embed the built UI into the go binary

https://github.com/AdguardTeam/AdGuardHome/blob/master/Makefile
https://github.com/AdguardTeam/AdGuardHome#how-to-build-from-source

Why do neither of these mention/use Packer? Why does this formula not use the `Makefile`?

Reply:
> Which installer? The node formula installs NPM to it's unclear why this would be different. What do you mean "doesn't work"? What is the failure?

`brew install`. When running with a `debug` flag and entering the shell, npm is not available, hence the npm install command fails.

> Why do neither of these mention/use Packer?

`packr` is used in the Makefile: https://github.com/AdguardTeam/AdGuardHome/blob/master/Makefile#L24

> Why does this formula not use the Makefile?

Because of the various env vars and paths used in the Makefile. It was easier to just build the binary and UI scripts in the brew context rather than do a wrapper for Makefile.

Reply:
> `brew install`. When running with a `debug` flag and entering the shell, npm is not available, hence the npm install command fails.

You can add to the path with `ENV.append_path "PATH", `

> `packr` is used in the Makefile: https://github.com/AdguardTeam/AdGuardHome/blob/master/Makefile#L24

Apologies, searched for `packer` and missed that.

> Because of the various env vars and paths used in the Makefile. It was easier to just build the binary and UI scripts in the brew context rather than do a wrapper for Makefile.

It would be more maintainable for us to use the `resource` to put `packr` in the right location but use the `Makefile` to do the other logic. This means if upstream changes the installation process we don't need to change anything/as much.

Thanks!

Reply:
How adding to a path solves the problem with npm being unavailable during build? Wouldn't that be a workaround for some problem with `node` formula if npm is available for `node@10` and `node@8` but not for `node` (v12)?

Moreover I just used an absolute path pointing to `/usr/local/bin/npm` and unfortunately the `fsevents` version used in AdGuard dependencies fail to build against Node.js v12.

> use the Makefile

If I understand correctly, formula requires downloading versioned source code archive which in case of GitHub is not a git repository. Because of that the first line of [Makefile will fail](https://github.com/AdguardTeam/AdGuardHome/blob/master/Makefile#L1).

Reply:
I'm not sure we should ship something that is dependent on old versions of node and doesn't properly define all the resources it needs to build. Is there an issue upstream for node 12/13 support? Why do we need to resource packr instead of it being a dependency?

As for the makefile, a lot of go projects have such a line. https://github.com/Homebrew/homebrew-core/issues/46083 uses the git download strategy for example. That would also solve that issue.

Reply:
> How adding to a path solves the problem with npm being unavailable during build? Wouldn't that be a workaround for some problem with `node` formula if npm is available for `node@10` and `node@8` but not for `node` (v12)?

No. The PATH is filtered differently because they are installed differently.

> Moreover I just used an absolute path pointing to `/usr/local/bin/npm` and unfortunately the `fsevents` version used in AdGuard dependencies fail to build against Node.js v12.

Ok, fine with me to use an older version for now.

> If I understand correctly, formula requires downloading versioned source code archive which in case of GitHub is not a git repository. Because of that the first line of [Makefile will fail](https://github.com/AdguardTeam/AdGuardHome/blob/master/Makefile#L1).

> As for the makefile, a lot of go projects have such a line. #46083 uses the git download strategy for example. That would also solve that issue.

Yup, this would work 👍 

> Why do we need to resource packr instead of it being a dependency?

@SMillerDev There's no current formula for it. It looks like the `Makefile` just downloads the latest version with `go get` which is non-ideal but we can use `go_resource` to handle this case (grep for `go_resource` usage in other formulae).

Reply:
@MikeMcQuaid  What about https://github.com/Homebrew/homebrew-core/issues/21801 regarding `go_resource`? Is it OK to use it?

Reply:
Turns out, that `fsevents` is a dev dependancy to `webpack` and it's a known issue that requires some updates. Because it's not a dependancy per se, those errors can be ignored and the client UI builds just fine.

I've pushed some changes. The formula now depends on `node`, uses the git download strategy and builds using Makefile. I left `packr` as a resource.